### DISCUSSION TITLE
the principle of ClusterData

### DISCUSSION THREAD

**Discussion by Chrisable** [pyg-team/pytorch_geometric#5194] [Q&A]
As I known,ClusterData uses the method of metis.But how to get the result of partition?
What does cluster_data.partptr  stand for？
I understand,  all_ node_index[partptr[0]:partptr[1]] is partition1, all_ node_index[partptr[1]:partptr[2]] is partition2.
If so,What is the difference between this and random node partitioning，If not,how to get the result of partition?

Here is my code:
```
import torch
from torch_geometric.datasets import Planetoid
from torch_geometric.transforms import NormalizeFeatures

dataset = Planetoid(root='data/Planetoid', name='PubMed', transform=NormalizeFeatures())
data = dataset[0]
from torch_geometric.loader import ClusterData, ClusterLoader
torch.manual_seed(12345)
cluster_data = ClusterData(data, num_parts=2)  # 1. Create subgraphs.
print("cluster_data.partptr:",cluster_data.partptr)
train_loader = ClusterLoader(cluster_data, batch_size=1, shuffle=True)
total_num_nodes = 0
for step, sub_data in enumerate(train_loader):
    print(f'Step {step + 1}:')
    print('=======')
    print(f'Number of nodes in the current batch: {sub_data.num_nodes}')
    print(sub_data.edge_index)
    total_num_nodes += sub_data.num_nodes

print(f'Iterated over {total_num_nodes} of {data.num_nodes} nodes!')
```

**Comment by rusty1s** [comment]
The easiest way to get the final permutation and partition is to add a node ID attribute to your data:
```python
data.n_id = torch.arange(data.num_nodes)
for subdata in ClusterData(data, …):
    print(subdata.n_id)

**Reply by Chrisable** [reply]
Is the result of each experiment of ClusterData the same？when i use` (edgecuts, parts) = metis.part_graph(G, 2)`,i found that the partition results of Metis and clusterdata are different

**Reply by rusty1s** [reply]
I think the METIS algorithm should be deterministic but not 100% sure TBH.

**Comment by Chrisable** [comment]
Thank you for your reply!By the way,What does cluster_data.partptr stand for？In my example,cluster_data.partptr: tensor([    0,  9870, 19717])

**Reply by rusty1s** [reply]
`partptr` denote the boundary indices of the clusters - that is: the first 9870 nodes belong to the first cluster, the nodes from 9870 to 19717 belong to the second cluster. This allows for efficient access of individual clusters - assuming that the graph has been permuted beforehand.

**Reply by Chrisable** [reply]
You mean：
```
data.n_id = torch.arange(data.num_nodes) 
cluster1 = data.n_id[0:9870] 
cluster2 = data.n_id[9870 :19717]
```
right?
But,What is the difference between this and random node partitioning?

**Reply by rusty1s** [reply]
Yes. The difference is that we use METIS to permute and partition the data - these intervals define the meaningful clusters in the graph, they are not random.

**Reply by Chrisable** [reply]
But i found: 
```
for cluster1:
sub_data.n_id: [   0    1 9882 ... 9878 9879 9881]
for cluster2:
sub_data.n_id: [    2     3     4 ... 14791 19715 19716]
```
cluster1 is not data.n_id[0:9870] because it exists 9878、 9879、 9881 and cluster2 is not data.n_id[9870 :19717]

**Reply by rusty1s** [reply]
Yes, the data is permuted such that the 9870 nodes in the first cluster are placed for in the `data`. That is why `sub_data.n_id` can contain values outside of 9870.

**Reply by Chrisable** [reply]
right?
![1660569900644](https://user-images.githubusercontent.com/49330390/184643636-7073a83b-83f3-486d-ba8c-79848b573850.jpg)

**Reply by rusty1s** [reply]
Yes :)

**Reply by karuna-bhaila** [reply]
Hi, just wanted to ask for some more clarification on this.
```
dataset = Planetoid(root='data/Planetoid', name='PubMed', transform=NormalizeFeatures())
data = dataset[0]
cluster_data = ClusterData(data, num_parts=2)
```
data.y contains the node labels in the original order that it was read in and cluster_data.data.y contains node labels in the new permuted order so that cluster_data.data.y[partptr[0]:partptr[1]] contains node labels for all nodes in the 1st cluster? Is this correct?

**Reply by rusty1s** [reply]
Yes, this is correct.

**Reply by karuna-bhaila** [reply]
Thank you for your reply
